packages/teleterm/src/ui/DocumentTerminal/useDocTerminal.ts

```diff
@@ -47,24 +48,50 @@ export default function useDocumentTerminal(doc: Props['doc']) {
   }
 
   function updateDocumentPid(pty: PtyProcess): void {
-    ctx.serviceDocs.update(doc.uri, { pid: pty.getPid() });
+    const pid = pty.getPid();
+    if (pid) {
+      ctx.serviceDocs.update(doc.uri, { pid });
+    }
+  }
+
+  function openTerminalContextMenu(): void {
+    ctx.mainProcessClient.openTerminalContextMenu();
   }
 
+  const refreshDocumentCwd = useMemo(() => {
+    return debounce(async () => {
+      const pid = ptyProcess.getPid();
+      const cwd = await ctx.serviceTerminals.getWorkingDirectory(pid);
+      ctx.serviceDocs.update(doc.uri, { cwd, title: cwd });
+    }, 2000);
+  }, [ptyProcess]);
+
   useEffect(() => {
+    let disposed = false;
     if (!ptyProcess) {
       createPtyProcess().then(pty => {
-        setPtyProcess(pty);
-        updateDocumentPid(pty);
+        if (!disposed) {
+          setPtyProcess(pty);
+          updateDocumentPid(pty);
+        }
       });
     }
 
+    ptyProcess?.onExit(() => {
+      const activeDocUri = ctx.serviceDocs.getActive().uri;
+      ctx.serviceDocs.close({ uri: activeDocUri });
+    });
+
     return () => {
+      disposed = true;
       ptyProcess?.dispose();
     };
   }, [ptyProcess]);
 
   return {
     ptyProcess,
+    refreshDocumentCwd,
+    openTerminalContextMenu,
```

Comment:
I wonder if we can make `ptyProcess` _not null_ during initial render? This is because this `null` introduces  multiple edge cases to handle. If I recall, it started from `getPreviousDocumentCwdIfPossible` so  maybe we can revisit this part and simplify?

For example, can a `doc` have `cwd` populated prior `useDocumentTerminal(doc)` is called? If we look at `openNewTerminal` it seems that it clones the `doc` object which might be what we need.

Reply:
Done as discussed